pkg/services/ngalert/state/template.go

```diff
@@ -0,0 +1,86 @@
+package state
+
+import (
+	"context"
+	"math"
+	"net/url"
+	"strconv"
+	"time"
+
+	text_template "text/template"
+
+	"github.com/grafana/grafana/pkg/services/ngalert/eval"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/template"
+)
+
+// templateCaptureValue represents each value in .Values in the annotations
+// and labels template.
+type templateCaptureValue struct {
+	Labels map[string]string
+	Value  float64
+}
+
+// String implements the Stringer interface to print the value of each RefID
+// in the template via {{ $values.A }} rather than {{ $values.A.Value }}.
+func (v templateCaptureValue) String() string {
+	return strconv.FormatFloat(v.Value, 'f', -1, 64)
+}
+
+func expandTemplate(name, text string, labels map[string]string, alertInstance eval.Result, externalURL *url.URL) (result string, resultErr error) {
+	name = "__alert_" + name
+	text = "{{- $labels := .Labels -}}{{- $values := .Values -}}{{- $value := .Value -}}" + text
+	data := struct {
+		Labels map[string]string
+		Values map[string]templateCaptureValue
+		Value  string
+	}{
+		Labels: labels,
+		Values: newTemplateCaptureValues(alertInstance.Values),
+		Value:  alertInstance.EvaluationString,
+	}
+
+	expander := template.NewTemplateExpander(
+		context.TODO(), // This context is only used with the `query()` function - which we don't support yet.
+		text,
+		name,
+		data,
+		model.Time(timestamp.FromTime(time.Now())),
```

Comment:
This is meant to be the evaluation time of the alert as opposed to `time.Now` - In the cache, you get that time via `			alertInstance.EvaluatedAt` and then pass it to this function.

Reply:
Thanks Josh, this is my mistake! I thought we could use `time.Now()` here.